locator/tablets.cc

```diff
@@ -585,6 +607,23 @@ class tablet_effective_replication_map : public effective_replication_map {
         return result;
     }
 
+    // FIXME: return a future object.
+    virtual dht::token_range_vector get_ranges(inet_address ep) const override {
```

Comment:
Where is this used?

Reply:
In `storage_service` class there is the following function. Previously, it was specific to `vnode_effective_replication_map` class. Now it uses `effective_replication_map`.

```
dht::token_range_vector         
storage_service::get_ranges_for_endpoint(const locator::effective_replication_map_ptr& erm, const gms::inet_address& ep) const {
    return erm->get_ranges(ep); 
}
```


It is used in a few places in `storage_service`.
One of such usages is in `storage_service::effective_ownership()`.

It looks as follows:

```
for (const auto& [dc, endpoints_map] : datacenter_endpoints) {
    for (auto endpoint : endpoints_map) {
        // calculate the ownership with replication and add the endpoint to the final ownership map
        try {
            float ownership = 0.0f;
            auto ranges = ss.get_ranges_for_endpoint(erm, endpoint);
            for (auto& r : ranges) {
                // get_ranges_for_endpoint will unwrap the first range.
                // With t0 t1 t2 t3, the first range (t3,t0] will be split
                // as (min,t0] and (t3,max]. Skippping the range (t3,max]
                // we will get the correct ownership number as if the first
                // range were not split.
                if (!r.end()) {
                    continue;
                }
                auto end_token = r.end()->value();
                auto loc = token_ownership.find(end_token);
                if (loc != token_ownership.end()) {
                    ownership += loc->second;
                }
            }
            final_ownership[endpoint] = ownership;
        }  catch (replica::no_such_keyspace&) {
            // In case ss.get_ranges_for_endpoint(keyspace_name, endpoint) is not found, just mark it as zero and continue
            final_ownership[endpoint] = 0;
        }
    }
}
```


I added `get_ranges(endpoint)` to `effective_replication_map` to force the derived classes to provide such functionality. Because of that I was able to reuse the logic in `storage_service::effective_ownership()` also for tablets.